pkg/cluster/deploy/deploy.go

```diff
@@ -0,0 +1,1803 @@
+package deploy
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/fatih/color"
+	"github.com/joomcode/errorx"
+	perrs "github.com/pingcap/errors"
+	"github.com/pingcap/tiup/pkg/cliutil"
+	"github.com/pingcap/tiup/pkg/cliutil/prepare"
+	"github.com/pingcap/tiup/pkg/cluster/clusterutil"
+	operator "github.com/pingcap/tiup/pkg/cluster/operation"
+	"github.com/pingcap/tiup/pkg/cluster/spec"
+	"github.com/pingcap/tiup/pkg/cluster/task"
+	"github.com/pingcap/tiup/pkg/errutil"
+	"github.com/pingcap/tiup/pkg/logger/log"
+	"github.com/pingcap/tiup/pkg/meta"
+	"github.com/pingcap/tiup/pkg/set"
+	"github.com/pingcap/tiup/pkg/utils"
+	"github.com/pingcap/tiup/pkg/version"
+	"golang.org/x/mod/semver"
+	"gopkg.in/yaml.v2"
+)
+
+//revive:disable:exported
+
+var (
+	errNSDeploy            = errorx.NewNamespace("deploy")
+	errDeployNameDuplicate = errNSDeploy.NewType("name_dup", errutil.ErrTraitPreCheck)
+)
+
+// Deployer to deploy a cluster.
+type Deployer struct {
+	sysName     string
+	specManager *spec.SpecManager
+}
+
+// NewDeployer create a Deployer.
+func NewDeployer(sysName string, specManager *spec.SpecManager) *Deployer {
+	return &Deployer{
+		sysName:     sysName,
+		specManager: specManager,
+	}
+}
+
+// StartCluster start the cluster with specified name.
+func (d *Deployer) StartCluster(name string, options operator.Options, fn ...func(b *task.Builder, metadata spec.Metadata)) error {
+	log.Infof("Starting cluster %s...", name)
+
+	metadata, err := d.meta(name)
+	if err != nil {
+		return perrs.AddStack(err)
+	}
+
+	topo := metadata.GetTopology()
+	base := metadata.GetBaseMeta()
+
+	b := task.NewBuilder().
+		SSHKeySet(
+			d.specManager.Path(name, "ssh", "id_rsa"),
+			d.specManager.Path(name, "ssh", "id_rsa.pub")).
+		ClusterSSH(topo, base.User, options.SSHTimeout).
+		Serial(task.NewFunc("StartCluster", func(ctx *task.Context) error {
+			return operator.Start(ctx, topo, options)
+		}))
+
+	if len(fn) > 0 {
+		if len(fn) != 1 {
```

Comment:
Can we support multiple callback functions?

Reply:
https://github.com/pingcap/tiup/pull/587/commits/51854efcf95ca02cc359d9ce4f19846edbc904a9